### PR TITLE
Play pattern

### PR DESCRIPTION
**Pull request by qchenevier**
In the `session` object, to complement the already existing `session.startRow` method, I propose a `session.play` method which would use a pattern language (a bit) similar to the one used for clips.

Several features added here:
- the `idx` parameter of a `Channel` can be a also string & several channels can have the same `idx`. 
  It's already the case today, you can specify any kind of `idx` and you're able to give several times the `idx`. However, the library wasn't designed explicitely to do so, it seems that the typescript code expects only (unique?) numbers.
  I propose to leverage that, in order to use this parameter as a kind of tag, accepting numbers and also strings, and modified the library to make it clear.  Maybe it would be relevant to rename this property/parameter and call it `label` or `tag`, rather than `idx`.
- introduction of a new pattern language, using the `Clip.idx` parameter to know which clip should be played and when. Each pattern is based on a '1-character per clip' principle, meaning that we can choose, for each time step to:
  - play a clip, numbered by an `idx` between 0 and 9. Example: to play 3 clips numbered in ascending order: `'123'`
  - extend a clip, using `'_'` as `idx` parameter. Example: to play this first clip during 3 time steps: `'1__'`
  - make some silence, using `'-'` as `idx` parameter. Example: wait 2 measure before starting: `'--1'`
- a method `session.play`, leveraging this pattern playing specification.

I think the example in  [script_play_pattern.js](https://github.com/qchenevier/scribbletune/blob/c278054c7af82150b9fd976c97647fe37e0fa634/dist/script_play_pattern.js), played through [`index_play_pattern.html`](https://github.com/qchenevier/scribbletune/blob/c278054c7af82150b9fd976c97647fe37e0fa634/dist/index_play_pattern.html), gives a good idea of what we can do with that feature.

That's an important change here, so you advice is wanted/recommended/welcome.

### CODE REVIEW

**Comment by walmik** [comment]
This is brilliant, i ll review in a day too @qchenevier

**Comment by qchenevier** [comment]
Thanks in advance. No rush, though. 🙂

The changes introduced here are quite impacting for the future & my intent with this PR is to discuss the new pattern language (like a "request for comments"). Having a first draft will help us have this discussion. That's "experimental code" so it's ok for me if we finally decide to have a different song pattern language than what is proposed here.

I'll ask also @mickmister to review/comment.

To anyone stumbling on this PR: feel free to review/comment also.

**Comment by qchenevier** [comment]
I noticed that I introduced a regression [here](https://github.com/qchenevier/scribbletune/commit/8a573c5caf0b47f0e1e499a87f5ddc8ad0819c9e#diff-f9c3b87ffde9e5d4256c51efff988cd5L262-R259) in an earlier PR. The last 3 commits are here to fix this (fix code, build, patch version bump).

If needed I can put it in another PR.

**Comment by mickmister** [comment]
@qchenevier Upon first inspection, I really like this. Especially the "section repeat" nature of the `_` operator. For the rest operator `-`, what's the use case for this? I'm thinking it would be for one "layer" or "instrument" to sit out for a bit, while some others are playing at that time?

@walmik Any way I can be added to assign myself as a reviewer for something like this?

**Comment by mickmister** [comment]
Also, I'm curious what it's like to do a midi export of a "song". Does each layer get exported in the same midi file, with the appropriate repeats and rests?

**Comment by walmik** [comment]
Thank you for offering to review @mickmister ! I ve added you as the reviewer... I m a bit swamped :/

File: dist/script_play_pattern.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      { pattern: '[-x][-x][-x]-', notes: 'C2' },
+    ],
+  });
+
+  window.session = session; // Useful only for debugging & API exploration, so that you can access the session object in the browser console
+
+  const channelPatterns = [
+    {
+      channelIdx: 'synth',
+      pattern: '-01_2_',
```
Comment (walmik) [inline dist/script_play_pattern.js:43]: This is awesome!

**Comment by walmik** [comment]
I think the `-` char is to "not play" the pattern for that duration. Something like Note off when `-` is used in a clip pattern. For the exported MIDI, it d be awesome to have the channels split up to MIDI channels. However I d imagine that s a feature by itself and maybe can be added separately than this PR.

**Comment by walmik** [comment]
Thank you @qchenevier I ll publish this now 👍

**Pull request merged**